plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/ConnectionWrapper.java

```diff
@@ -0,0 +1,244 @@
+package org.jboss.tools.openshift.internal.ui.models;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.OperationCanceledException;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.jobs.Job;
+import org.jboss.tools.openshift.core.connection.ConnectionProperties;
+import org.jboss.tools.openshift.core.connection.IOpenShiftConnection;
+import org.jboss.tools.openshift.internal.core.WatchManager;
+import org.jboss.tools.openshift.internal.ui.OpenShiftUIActivator;
+
+import com.openshift.restclient.ResourceKind;
+import com.openshift.restclient.model.IProject;
+import com.openshift.restclient.model.IResource;
+
+public class ConnectionWrapper extends AbstractOpenshiftUIElement<IOpenShiftConnection, OpenshiftUIModel> {
+	public static final String[] RESOURCE_KINDS = { ResourceKind.BUILD, ResourceKind.BUILD_CONFIG,
+			ResourceKind.DEPLOYMENT_CONFIG, ResourceKind.IMAGE_STREAM, ResourceKind.IMAGE_STREAM_TAG, ResourceKind.POD,
+			ResourceKind.ROUTE, ResourceKind.REPLICATION_CONTROLLER, ResourceKind.SERVICE };
+
+	private AtomicReference<LoadingState> state = new AtomicReference<LoadingState>(LoadingState.INIT);
+	private Map<IProject, ProjectWrapper> projects = new HashMap<>();
+	private ResourceCache resourceCache = new ResourceCache();
+
+	public ConnectionWrapper(OpenshiftUIModel parent, IOpenShiftConnection wrapped) {
+		super(parent, wrapped);
+	}
+
+	public IOpenShiftConnection getConnection() {
+		return getWrapped();
+	}
+
+	public Collection<ProjectWrapper> getProjects() {
+		synchronized (projects) {
+			return new ArrayList<ProjectWrapper>(projects.values());
+		}
+	}
+
+	public LoadingState getState() {
+		return state.get();
+	}
+	
+	void initWith(List<IResource> resources) {
+		synchronized (projects) {
+			resources.forEach(project -> {
+				projects.put((IProject) project, new ProjectWrapper(this, (IProject) project));
+			});
+			state.set(LoadingState.LOADED);
+		}
+	}
+
+	public boolean load(IExceptionHandler handler) {
+		if (state.compareAndSet(LoadingState.INIT, LoadingState.LOADING)) {
+			startLoadJob(handler);
+			return true;
+		}
+		return false;
+	}
+
+	void startLoadJob(ProjectWrapper projectWrapper, IExceptionHandler handler) {
+		new Job("load project") {
+
+			@Override
+			protected IStatus run(IProgressMonitor monitor) {
+				try {
+					WatchManager.getInstance().startWatch(projectWrapper.getResource());
+					Collection<IResource> resources = new HashSet<>();
+					for (String kind : RESOURCE_KINDS) {
+						resources.addAll(projectWrapper.getResource().getResources(kind));

```

Comment:
I don't think we did this previously, but we should be retrieving resources from the connection instead of the project.  This ensures we properly handle the auth logic and token expiration.